### PR TITLE
reframing data pipelines

### PR DESCRIPTION
**Pull request by dimitri-yatsenko**

### CODE REVIEW

File: datajoint_workflows.md [hunk 1]
```diff
@@ -1,51 +1,46 @@
...
-## Terms
+**Authors**: Raphael Guzman and Dimitri Yatsenko
 
-Before we get too ahead of ourselves, it might be best to establish a set of definitions to help guide the message presented here so we can agree on a common language.
+## Definition 
+Let's agree on a few  definitions:
 
-*a **job** is a fully-specified request to perform a computation or task in a deterministic manner*
+### Jobs
+A **job** is a fully-specified request to perform a computation or task in a deterministic manner
```
Comment (guzman-raphael) [inline datajoint_workflows.md:22]: Prefer if we highlight or quote this line is some way as it is the concise description of what a job is and the lasting impression of what they should take away from this section. I've noticed this pattern quite a bit in blogs I've come across before.

```suggestion
*A **job** is a fully-specified request to perform a computation or task in a deterministic manner.*
```

File: datajoint_workflows.md [hunk 2]
```diff
@@ -1,51 +1,46 @@
...
 
-The idea is that jobs should be specified in their entirety such that they can achieve consistent results in a predictable timeframe. Reliability is achieved if results can be reproduced and independently verifed.
+Jobs should be specified in their entirety so that they can achieve consistent results in a predictable timeframe. Reliability is achieved if results can be reproduced and independently verifed.
 
-For instance, let's say I have a job that is responsible for compiling a `plotly` graph based on some input data. To reliability achieve similar results, there are several questions we need to pin down. To name a few:
+Consider a job that responsible for compiling a `plotly` graph from some input data. For reliable results, we need to answer several questions:
```
Comment (guzman-raphael) [inline datajoint_workflows.md:26]:
```suggestion
Consider a job that is responsible for compiling a `plotly` graph from some input data. For reliable results, we need to answer several questions:
```

**Comment by ttngu207** [comment]
I have some thoughts/comments on the terminology here, redefining some of the terms

+ ***task***: a single unit of analysis/computation - minimally requires
    + input
    + output
    + software packages, dependencies
    + hardware configuration requirements
+ ***workflow***: a sequence of ***task*** (optionally) chained together to perform some meaningful process
+ ***job***: an instance of ***workflow*** being created, with a set of input, queued, scheduled to be executed at a certain datetime, on a certain computing resource, using a certain software image

***job orchestration*** refers to the overall processing of creating, managing, monitoring, and interacting with these ***jobs***

***workflow management*** is on a higher level, managing:
+ tasks
+ workflows
+ computing resources
+ software images
+ jobs and job orchestration

**Comment by dimitri-yatsenko** [comment]
Also, I think we should limit this article to workflow management, leaving other topics such as query performance and schema migration for a separate article.

**Comment by ttngu207** [comment]
***DataJoint*** is a pipeline definition framework. What is DataJoint pipeline? DataJoint pipelines define the ***tasks*** and ***workflows*** according to the definition above, with the added benefit of ***data*** being a first-class citizen of every ***task*** (with built-in serialization for scientific data!).

+ DataJoint table = ***task*** + ***data***
+ DataJoint sub-DAG = ***workflow***
+ a `key` to be populated is a ***job*** (not quite yet, a `key` would only defines the input, ***task*** and ***workflow***, not yet software env and hardware config)

The ***DataJoint*** framework provides a lightweight, albeit rudimentary `job orchestration` solution, via the `.populate(reserve_jobs=True)` mechanism

***DataJoint ecosystem*** does not have a full solution for ***workflow management*** (per its definition above)

File: datajoint_workflows.md [hunk 3]
```diff
@@ -1,51 +1,46 @@
...
+- What version of `plotly` and all its dependencies will we use?
+- What is the input data and the processing parameters?
+- What hardware resources should we allocate? CPU? GPU? RAM? What operating system and hardware architecture?
+- How can the job be parallelized through multithreading and multiprocessing?
 
-Though some of these last questions may not affect the final result, they can affect the duration in which a job is performed. By capturing these properly, we can provide with greater confidence estimates of both the expected results and duration.
+These last questions may not affect the final result but they do affect the run time.  Answering these questions leads to the predictability of computation results and execution times.
 
-*a **workflow** defines the acyclic relationship of many jobs to each other*
+A **workflow** is the formal specification of the dependency relationship between jobs
```
Comment (guzman-raphael) [inline datajoint_workflows.md:35]: Curious on why you feel we should remove the `acyclic` description. Think this is an important definition as workflows must terminate. Also, similarly as above.

```suggestion
### Workflows
*A **workflow** is the formal specification of the acyclic relationship between jobs.*
```

File: datajoint_workflows.md [hunk 4]
```diff
@@ -1,51 +1,46 @@
...
 
-- embed the graph within a provided text input (say for a report)
+- embed the plot within a provided text input (say for a report)
 - log the activity to a central service
 
-Here we've described a workflow with one parent job and two children jobs.
+This workflow has one parent job and two children jobs.
 
-***job orchestration** is the process of delegating jobs to both an available and appropriate computing resource*
+Then **job orchestration** is the process of delegating jobs to both an available and appropriate computing resource.
```
Comment (guzman-raphael) [inline datajoint_workflows.md:46]: Similarly as above.

```suggestion
### Job Orchestration
***Job orchestration** is the process of delegating jobs to both an available and appropriate computing resource.*
```

File: datajoint_workflows.md [hunk 5]
```diff
@@ -1,51 +1,46 @@
...
+Think of the job orchestrator layer as a matchmaker of sorts. By listening to a *job queue*, her responsibility is to connect the right job to the right environment; best fit for the job. Notice that the notion of a workflow here really isn't necessary since all we are doing is scheduling and fullfilling computing obligations. This means that as we process the queue, we either need to spawn an environment suitable for the job or wait until there is bandwidth within the landscape (remote or on-premise) to launch it.
 
 Perhaps in our example we'd like for our jobs to be run within a distributed Kubernetes cluster. Jobs could be encapsulated in pods scheduled against nodes that satisfy the hardware requirements.
 
-***workflow management** is an interface (graphical or programatic) that enables workflow specification and job queueing*
+**Workflow Management** is an interface (graphical or programatic) that enables workflow specification and job queueing
```
Comment (guzman-raphael) [inline datajoint_workflows.md:52]: Similarly as above.

```suggestion
### Workflow Management
***Workflow management** is an interface (graphical or programatic) that enables workflow specification and job queueing.*
```

**Comment by ttngu207** [comment]
The terms ***DataJoint pipeline*** and ***DataJoint workflow*** have been used somewhat interchangeably. Let's add specific definition and distinctions.

DataJoint pipeline is a ***workflow*** in the sense that it defines the ***task*** (in the form of table) and the sequence/dependencies (in the form of relational data model). DataJoint pipeline takes one step further to define the data and the input/output format at each ***task*** (in the form of the table attributes). In fact, a DataJoint pipeline captures a collection of ***workflows*** and their dependencies. 

Ones can argue that DataJoint pipeline is not a ***workflow*** for those reasons as well: ***workflow*** may just be chained computational steps without the needs of any accompanying data. The tightly coupled data and code in DataJoint's philosophy could mean that for those looking to use DataJoint purely as a workflow, it may not well suited. 
 
Perhaps DataJoint data pipeline can be thought of as:
 > relational database with embedded code
 
 or
 
 > workflow with embedded database

File: datajoint_workflows.md [hunk 6]
```diff
@@ -55,7 +50,10 @@ Naturally, users will need to interact with these facilities to perform work. Th
 
 Returing to our example, using a workflow management tool would be how we might define our workflow's jobs and set the workflow to be triggered once a day. Though largely automated now, we'd be wise if during processing we assign someone to be on-call should intervention be necessary.
 
-*a **data pipeline** establishes the acyclic relationship between workflows separated only by necessary input parameters, provided by users, achieving a complete goal*
+Now a **data pipeline** is somewhat different conceptually.  
+It is the formal specification of the data sources, transformations, and all its intermediate structures.
+It also takes the form of a directed acyclic graph of dependencies. 
+In many cases, the data pipeline provides the basis for the *workflows*.
```
Comment (guzman-raphael) [inline datajoint_workflows.md]: This is too long. We need a short, concise description of a **data pipeline** that can easily be quoted from this blog. Also, similarly as above. I don't think this captures that data

```suggestion
### Data Pipeline
*A **data pipeline** is the formal specification of the data sources, transformations, and all its intermediate structures that provide the basis for many workflows.*
```

File: datajoint_workflows.md [hunk 7]
```diff
@@ -69,9 +67,9 @@ To complete our example, our larger effort might be that we are conducting a beh
 - workflow: add pose tracking to videos and update metadata, input: utilize a specific algorithm
 - workflow: generate graphs and figures for publication, input: curated list of session data
 
-## DataJoint OSS' Strengths
+## DataJoint's Approach
 
-When we started the DataJoint open-source framework back in 2011, there were several key principles we wanted to adhere to that we felt were either missing or reduced in significance in the management of computation at the time; primarily within neuroscience. Our approach is to empower the user to create and manage relational **data pipelines**. If you are not familiar with the DataJoint open-source framework, we ecourage you to go and check out [this](https://datajoint.org or initial paper) first.
+When we designed the DataJoint framework, there were several key principles we wanted to adhere to that we felt were either missing or reduced in significance in the management of computation at the time; primarily within neuroscience. Our approach is to empower the user to create and manage relational **data pipelines**. If you are not familiar with the DataJoint framework, please see [this](https://datajoint.org) or initial paper.
```
Comment (guzman-raphael) [inline datajoint_workflows.md]:
```suggestion
When we designed the DataJoint framework, there were several key principles we wanted to adhere to that we felt were either missing or reduced in significance in the management of computation at the time; primarily within neuroscience. Our approach is to empower the user to create and manage relational **data pipelines**. If you are not familiar with the DataJoint framework, please see [https://datajoint.org](https://datajoint.org) and our original [position paper](https://www.biorxiv.org/content/10.1101/031658v1).
```

File: datajoint_workflows.md [hunk 8]
```diff
@@ -1,51 +1,46 @@
...
 **Duration Target:** 14min read (@160 wpm)
 
 **Release:** Thursday, 2022-03-24
 
-**Title:** Blog Title
-
-## [DRAFT] Intro
-- something clever/attention-grabbing
-- soften tone to be open-minded as an opinionated discussion follows
-
```
Comment (guzman-raphael) [inline datajoint_workflows.md]: We still need this section introduce the discussion. These are notes for consideration. I'm currently working on this and will be updated soon.

**Comment by ttngu207** [comment]
***DataJoint*** introduces notion of `dj.Manual`, `dj.Imported`, `dj.Computed` to provide rich language to define a pipeline.
+ a `dj.Manual` table specifies a point requesting manual user-input - essentially providing the input, a `key` to the immediate downstream ***workflow***
+ a `dj.Imported` or `dj.Computed` table right downstream from a `dj.Manual` table is a natural starting point of a new ***workflow*** (as defined above). Going with this thought process, then the `dj.Imported` or `dj.Computed` immediate above a `dj.Manual` table could be a natural stopping point of a ***workflow*** as well

File: datajoint_workflows.md [hunk 9]
```diff
@@ -1,51 +1,45 @@
+# Workflow Management for DataJoint
```
Comment (guzman-raphael) [inline datajoint_workflows.md:1]: What about?

```suggestion
# DataJoint-driven Workflow Management
```

File: datajoint_workflows.md [hunk 10]
```diff
@@ -55,7 +49,10 @@ Naturally, users will need to interact with these facilities to perform work. Th
 
 Returing to our example, using a workflow management tool would be how we might define our workflow's jobs and set the workflow to be triggered once a day. Though largely automated now, we'd be wise if during processing we assign someone to be on-call should intervention be necessary.
 
-*a **data pipeline** establishes the acyclic relationship between workflows separated only by necessary input parameters, provided by users, achieving a complete goal*
+A **data pipeline** is somet different conceptually.  
+It is the formal specification of the data sources, transformations, and all its intermediate structures.
+It also takes the form of a directed acyclic graph of dependencies. 
+In many cases, the data pipeline provides the basis for the *workflows*.
```
Comment (guzman-raphael) [inline datajoint_workflows.md]: This is too long. We need a short, concise description of a data pipeline that can easily be quoted from this blog. Also, similarly as above. I don't think this captures that data

```suggestion
*A **data pipeline** is the formal specification of the data sources, transformations, and all its intermediate structures that provide the basis for many workflows.*
```

File: datajoint_workflows.md [hunk 11]
```diff
@@ -1,51 +1,45 @@
...
+
 **Duration Target:** 14min read (@160 wpm)
 
 **Release:** Thursday, 2022-03-24
 
-**Title:** Blog Title
-
-## [DRAFT] Intro
-- something clever/attention-grabbing
-- soften tone to be open-minded as an opinionated discussion follows
```
Comment (guzman-raphael) [inline datajoint_workflows.md:9]: Since we are still working on adding this, I think we should keep these notes until the section is completed.

**Review by guzman-raphael — APPROVED** [review]
Let's continue discussing and collaborating on this. Merging this and will follow-up with additional contributions.

**Pull request merged**